src/services/hash.password.bcryptjs.ts

```diff
@@ -0,0 +1,19 @@
+import {genSalt, hash} from 'bcryptjs';
+
+/**
+ * Service HashPassword using module 'bcryptjs'.
+ * It takes in a plain password, generates a salt with given
+ * round and returns the hashed password as a string
+ */
+export type HashPassword = (
+  password: string,
+  rounds: number,
```

Comment:
IMO, consumers of `HashPassword` should not need to understand specifics of the hashing function used, e.g. number of rounds. Typically, we want to use a different number of rounds when running in test vs. in production. Consumers of `HashPassword` do not have enough information to be able to pick the right number of rounds.

I am proposing the following API:

```ts
export type HashPassword = (password: string) => Promise<string>;
```

And the following usage:

```ts
// in production
app.bind(OtherServicesBindings.HASH_PASSWORD).to(pwd => hasPassword(pwd, 10));

// in test
app.bind(OtherServicesBindings.HASH_PASSWORD).to(pwd => hasPassword(pwd, 4));
```

Having wrote that, I think we should go one step further. As I see it, the function for computing password hash is tightly coupled with the function for comparing user-provided password with the stored hash, and therefore both functions should be part of the same "service".

```ts
interface PasswordHasher {
  hash(password: string): Promise<string>;
  compare(password: string): Promise<boolean>;
}

class BcryptHasher implements PasswordHasher {
  constructor(
    @inject('BcryptHasher.rounds')
    private readonly rounds: number
  ) {}
  // ...
}

app.bind('services.PasswordHasher).toClass(BcryptHasher);

// usage in production
app.bind('BcryptHasher.rounds').to(10);
// usage in tests
app.bind('BcryptHasher.rounds').to(4);
```

Later in the future we should leverage https://github.com/strongloop/loopback-next/pull/2259 to control this configuration option.

Feel free to leave this second part (service for hash+compare) for a follow-up pull request.

Reply:
@bajtos 
> IMO, consumers of HashPassword should not need to understand specifics of the hashing function used, e.g. number of rounds.

> Having wrote that, I think we should go one step further. As I see it, the function for computing password hash is tightly coupled with the function for comparing user-provided password with the stored hash, and therefore both functions should be part of the same "service".


Both good points 👍 Do you mind if I address both of them in the next PR as a refactor specific for password hashing? (within the same story https://github.com/strongloop/loopback4-example-shopping/issues/40)

I would like to have this PR focus on the JWT related services first. Then improve the password hash in a second one.

Reply:
Fair enough 👍